File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Evaluation/Strategies/ApplicationHostnameStrategy.cs
namespace Gatrix.Server.Sdk.Evaluation.Strategies;

/// <summary>
/// Application hostname strategy — enables for specific hostnames.
/// </summary>
public class ApplicationHostnameStrategy : IStrategy
{
    public string Name => "applicationHostname";

    public bool IsEnabled(Models.StrategyParameters? parameters, Models.EvaluationContext context, string flagName)
    {
        var hostNames = parameters?.HostNames;
        if (string.IsNullOrEmpty(hostNames)) return false;

        var currentHostname = System.Net.Dns.GetHostName();
        var hosts = hostNames.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
        return hosts.Contains(currentHostname, StringComparer.OrdinalIgnoreCase);
    }
}

File: packages/evaluator/src/strategies/application-hostname-strategy.ts
/**
 * Application Hostname Strategy
 *
 * Enables feature for applications running on specific hosts.
 */

import { hostname } from 'os';
import {
  EvaluationContext,
  StrategyParameters,
  StrategyEvaluationResult,
} from '@gatrix/shared';
import { Strategy } from './strategy';

export class ApplicationHostnameStrategy extends Strategy {
  private readonly currentHostname: string;

  constructor() {
    super('applicationHostname');
    this.currentHostname = (
      process.env.HOSTNAME ||
      hostname() ||
      'undefined'
    ).toLowerCase();
  }

  isEnabled(
    parameters: StrategyParameters,
    _context: EvaluationContext
  ): boolean {
    const hostNames = parameters.hostNames;
    if (!hostNames) return false;
    const matched = hostNames
      .toLowerCase()
      .split(/\s*,\s*/)
      .includes(this.currentHostname);
    return matched;
  }

  isEnabledWithDetails(
    parameters: StrategyParameters,
    _context: EvaluationContext
  ): StrategyEvaluationResult {
    const hostNames = parameters.hostNames;
    if (!hostNames) {
      return { enabled: false, reason: 'No hostNames defined in parameters' };
    }

    const hostList = hostNames.toLowerCase().split(/\s*,\s*/);
    const matched = hostList.includes(this.currentHostname);

    return {
      enabled: matched,
      reason: matched
        ? `Hostname "${this.currentHostname}" found in list`
        : `Hostname "${this.currentHostname}" not in list [${hostList.join(', ')}]`,
      details: { currentHostname: this.currentHostname, hostList },
    };
  }
}
